Sprint 2/step 2

ArtemixxPG opened the pull request.

ArtemixxPG committed:
add custom hook useIngredients for context in burger-constructor

ArtemixxPG committed:
add custom hook useIngredients for context in burger-constructor

ArtemixxPG committed:
add POST for order
add dinamical counter sum order

ArtemixxPG committed:
add POST for order
add dinamical counter sum order

ArtemixxPG committed:
add destructuring props
add folder 'services' for context
remove JSON.stringify from deps useEffects
use const from let

ArtemixxPG committed:
remove additional `props` from project

ArtemixxPG committed:
change calc unique key for constructor burger

ArtemixxPG committed:
change ror sprint-2/step-2

ArtemixxPG committed:
Merge remote-tracking branch 'origin/sprint-2/step-1' into sprint-2/step-2

# Conflicts:
#	react-burger/src/componets/burger-constructor/burger-constructor.jsx
#	react-burger/src/componets/home/home.jsx
#	react-burger/src/componets/modal-content/modal-content-order-complete/modal-content-order-complete.jsx
#	react-burger/src/custom-hooks/use-fetch-list.jsx
#	react-burger/src/utils/URL.js
#	react-burger/src/utils/utils.js

vadmas commented on react-burger/src/componets/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,63 +1,128 @@
 
-    const [open, setOpen] = useState(false)
 
+     const [open, setOpen] = useState(false)
```
Нужно исправить: данный хук избыточен. Модальное окно следует открывать, если есть номер заказа (order.number).  А запрос отправлять при нажатии на "оформить заказ". Промежуточное состояние здесь не стоит использовать

vadmas commented on react-burger/src/componets/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,63 +1,128 @@
-        )
+    const [failContentModal, setFailContentModal] = useState(false)
+
+    const {buns, sauces, mains} = useSelector (store => store.ingredients.types)
```
Отлично: правильно реализована работа с хуком useSelector

vadmas commented on react-burger/src/componets/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,63 +1,128 @@
+
+
+
+     useEffect(()=>{
```
Нужно исправить: здесь нужно использовать useMemo, а не dispatch + useEffect

vadmas commented on react-burger/src/componets/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,63 +1,128 @@
+     const setIngredient = (id) => {
+         const ingredients = [...buns, ...sauces, ...mains]
+         const ingredient = ingredients.find(item=> item._id === id.id)
+         ingredient.type === 'bun' ? dispatch({type: ADD_BUN, payload: ingredient})
```
Нужно исправить: уместнее использовать значение из констант INGREDIENT_TYPES.BUNS

vadmas commented on react-burger/src/componets/burger-constructor/burger-cunstructor-item/burger-constructor-item.jsx:
```diff
@@ -0,0 +1,89 @@
+import {REMOVE_INGREDIENT, SET_INGREDIENTS} from "../../../services/actions/selected-ingedients-actions";
+import {sortArray} from "../../../utils/utils";
+
+const BurgerConstructorItem = ({name, image, price, index, id}) => {
```
Нужно исправить: нет валидации пропсов. Стоит описать типы пропсов данного компонента, используя библиотеку prop-types

vadmas commented on react-burger/src/componets/burger-constructor/burger-cunstructor-item/burger-constructor-item.jsx:
```diff
@@ -0,0 +1,89 @@
+
+
+
+    const [{ handlerId }, dropItem] = useDrop({
```
Отлично: хорошо реализована работа с react-dnd

vadmas commented on react-burger/src/componets/burger-ingredient/burger-ingredient.jsx:
```diff
@@ -46,7 +49,7 @@ const BurgerIngredient = (props) => {
 BurgerIngredient.propTypes = {
     buns: PropTypes.arrayOf(IngredientPropTypes),
```
Нужно исправить: здесь правильно добавить ещё один isRequired: PropTypes.arrayOf(IngredientPropTypes.isRequired).isRequired
Так же касается случаев ниже

vadmas commented on react-burger/src/services/actions/ingedients-actions.js:
```diff
@@ -0,0 +1,24 @@
+export const SUCCESS_REQUEST_INGREDIENTS = 'SET_INGREDIENTS'
+export const ERROR_REQUEST_INGREDIENTS = 'ERROR_REQUEST_INGREDIENTS'
+
+export const getIngredients = () => dispatch => {
```
Отлично: правильно реализована работа с санками (redux-thunk)

vadmas commented on react-burger/src/services/actions/ingedients-actions.js:
```diff
@@ -0,0 +1,24 @@
+    dispatch({type: REQUEST_INGREDIENTS})
+
+    fetch(URL_GET_INGREDIENTS)
+        .then(res => res.json())
```
Нужно исправить: при таком запросе, если он завершится ошибкой (статус не 200),
то колбэк всеравно будет вызван и в стейт запишется undefined.
Поэтому следует сделать дополнительно проверку успешности выполнения запроса res.ok ,
например, return res.ok ? res.json() : res.json().then((err) => Promise.reject(err)).
Такую проверку стоит вынести в отдельную функцию, в дальнейшем она будет переиспользоваться

vadmas commented on react-burger/src/services/actions/ingedients-actions.js:
```diff
@@ -0,0 +1,24 @@
+            if (data.success) {
+                return data.data
+            } else {
+                dispatch({type: ERROR_REQUEST_INGREDIENTS})
```
Можно лучше: в данном случае, дабы не дублировать вызов экшена, уместнее возвращать Promise.reject("текст ошибки").  А в блоке catch уже единообразно будет вызываться экшен, в случае возникшей ранее ошибки

vadmas commented on react-burger/src/services/reducers/order-reducer.js:
```diff
@@ -0,0 +1,43 @@
+                hasLoading: true
+            }
+        case SUCCESS_REQUEST_ORDER:
+            console.log(action.payload)
```
Можно лучше: оставлять `console.log` в коде после завершения отладки считается плохим тоном

vadmas commented on react-burger/src/services/reducers/selected-ingedients-reducers.js:
```diff
@@ -0,0 +1,52 @@
+const initialState = {
+    selectedBun: null,
+    selectedIngredients: [],
+    totalPrice: 0
```
Можно лучше: полную стоимость лучше рассчитывать в селекторе (на стороне компонента), а не хранить в сторе

vadmas commented on react-burger/src/index.tsx:
```diff
@@ -3,13 +3,35 @@ import ReactDOM from 'react-dom/client';
+
+const enhancer = composeEnhancers(applyMiddleware(thunk));
+
+const store = configureStore({reducer: rootReducer, enhancers:[enhancer]});
```
Нужно исправить: если вы используете configureStore, то composeEnhancers, applyMiddleware - становятся избыточными.  Правильнее описать стор так: const store = configureStore({reducer: rootReducer, middleware: [thunk], devTools: true})

vadmas reviewed (commented):
Хорошая работа.
Приложение приобретает готовый вид, видно как вы прогрессируете по мере прохождения курса.
В положительную сторону хочется отметить:
- Хорошо реализована работа с redux-thunk
- Редъюсеры - чистые функции, разделены по сущностям,
- В рамках одного описания экшена выполнятся преобразование состояния стора - отлично!
- Взаимодействие сторов с компонентами реализовано через хуки useSelector, useDispatch
- Работы с redux-dnd реализована отлично - молодец!
- Реализована навигация по разделам ингредиентов - отлично!
- Функционал конструктора бургера реализован верно

**Но есть несколько замечаний:**

**Замечания которые необходимо будет исправить:**
- Небольшие недочеты в работе с запросами (смотри комментарий в ingedients-actions.js, getIngredients).
  Эти же замечания актуальны и для order-actions.js, orderPost

Так же стоит учесть комментарии, оставленные в коде.

**Места, в которых можно улучшить проект:**
- Стоит обращать внимания на предупреждения в консоли и терминале запущенного приложения. Это хорошие рекомендации, которых стоит придерживаться для улучшения структуры кода и качества его написания


Пожалуйста, учтите выше описанные замечания и те, что оставлены коде вашей работы (особенно с пометкой "нужно исправить").
При повторной проверке, если все замечания будут исправлены, работа будет принята.

Удачного рефакторинга кода!

ArtemixxPG committed:
first change review

ArtemixxPG committed:
change ror sprint-2/step-2:
add Preloader
remove console.log
change open modal in burger constructor
change response => response ok
remove total sum => calculate total sum in client
change configure store
add proptypes in burger constructor item
remove useEffect => add useMemo

vadmas commented:
Отлично, все критичные комментарии исправлены. Работу принимаю. Удачи в дальнейшем обучении!

ArtemixxPG committed:
add auth page
